### ISSUE TITLE
Support SFT

### ISSUE THREAD

**Issue by rlrs** [rlrs/maester#20]
# SFT Design for Maester

## Goal
Add minimal supervised fine-tuning (SFT) support to Maester for training on conversation data.

### 1. Data Format
- **Input**: Parquet files with a `conversations` column containing JSON arrays
- **Format**: `[{"role": "user", "content": "..."}, {"role": "assistant", "content": "..."}]`
- **Output**: Dict with `input_ids`, `labels`, `attention_mask` for easy extension

### 2. Architecture
```
maester/sft/
├── messages.py      # Message class with role enum, masking strategies
├── templates.py     # ChatML template (no Jinja2 for now)
├── tokenization.py  # Convert messages → tokens with masking
└── dataset.py       # ConversationParquetDataset extends ParquetDataset
```

### 3. Key Components
- **Message class**: Simple dataclass with `role`, `content`, `masked`, `eot` fields
- **Masking**: Only train on assistant responses by default (labels = -100 for masked tokens)
- **No packing**: One conversation per sequence, right-padded to `max_seq_len` (or minimum of the maximum of the sequence lengths of the samples in the batch and max_seq_len)?
- **Template**: Direct string formatting, not Jinja2 (simpler, no dependencies)

### 4. Integration Points
```python
# In train.py (minimal changes)
if cfg.sft.enabled:
    dataloader = build_sft_data_loader(cfg, rank, world_size)
else:
    dataloader = build_experimental_data_loader(cfg, rank, world_size)
```

### 5. Future Extensions
- **Distillation**: Dict outputs allow adding `teacher_logits` field later
- **Packing**: Can add conversation packing for efficiency
- **Templates**: Easy to add Llama3, Alpaca formats

## Benefits
- Reuses all existing infrastructure (checkpointing, distributed training)
- Not too many lines of code
- No or few modifications to core training loop
- Clear upgrade path for further features

## Definition of Done
- Can load conversation data from Parquet files
- Correctly masks user/system messages in loss computation
- Works with existing distributed training setup
- Checkpointing/resumption works correctly
- Simple example demonstrating usage

## Why Merge This?
1. **Immediate value**: Enables fine-tuning without switching frameworks
2. **Low risk**: Minimal changes to existing code, feature is optional
3. **Foundation for more**: Clean base for distillation, RLHF, and other post-training methods
4. **Maintains philosophy**: Simple, explicit, debuggable - no hidden complexity
